File: automation/clean-data.mjs
import fs from "node:fs/promises";
import path from "node:path";
import XLSX from "xlsx";

const args = parseArgs(process.argv.slice(2));

if (!args.orders || !args.routes || !args.config) {
  console.error("用法：node automation/clean-data.mjs --orders 订单列表.xlsx --routes 销转表.xlsx --config automation/config.json");
  process.exit(1);
}

const outputDir = path.resolve(args["output-dir"] || "automation/outputs");
await fs.mkdir(outputDir, { recursive: true });

const config = JSON.parse(await fs.readFile(path.resolve(args.config), "utf8"));
const orderRows = readSheetRows(args.orders);
const routeRows = readRouteExport(args.routes);
const results = [];

for (const task of config.tasks.filter((item) => item.enabled)) {
  const routeConfig = findRouteConfig(config.routes, task.routeName);
  if (!routeConfig) {
    throw new Error(`未匹配总路线配置：${task.routeName}`);
  }

  const taskOrders = filterOrders(orderRows, task);
  if (taskOrders.length === 0) {
    results.push({ task, status: "empty", message: "没有有效已付款旅客" });
    continue;
  }

  const routeInfo = matchRouteInfo(routeRows, taskOrders, task);
  const payload = buildInsurancePayload(task, routeConfig, routeInfo, taskOrders);
  const baseName = safeFilename(`${task.routeName}_${task.startDate}`);
  const jsonPath = path.join(outputDir, `${baseName}.json`);
  const txtPath = path.join(outputDir, `${baseName}_粘贴名单.txt`);

  await fs.writeFile(jsonPath, JSON.stringify(payload, null, 2), "utf8");
  await fs.writeFile(txtPath, payload.pasteList, "utf8");
  results.push({
    task,
    status: "ok",
    travelerCount: payload.travelers.length,
    json: jsonPath,
    pasteList: txtPath,
  });
}

const summaryPath = path.join(outputDir, "summary.json");
await fs.writeFile(summaryPath, JSON.stringify(results, null, 2), "utf8");
console.log(JSON.stringify({ summary: summaryPath, results }, null, 2));

function readSheetRows(filePath) {
  const workbook = XLSX.readFile(path.resolve(filePath), { cellDates: false });
  const sheet = workbook.Sheets[workbook.SheetNames[0]];
  return XLSX.utils.sheet_to_json(sheet, { defval: "", raw: false });
}

function readRouteExport(filePath) {
  const workbook = XLSX.readFile(path.resolve(filePath), { cellDates: false });
  const sheet = workbook.Sheets[workbook.SheetNames[0]];
  const matrix = XLSX.utils.sheet_to_json(sheet, { header: 1, defval: "", raw: false });
  if (matrix.length < 2) throw new Error("销转表格式异常：少于两行表头");
  const headers = matrix[1].map((value) => String(value || "").trim());
  return matrix.slice(2).map((row) => {
    const item = {};
    headers.forEach((header, index) => {
      item[header] = row[index] ?? "";
    });
    return item;
  });
}

function filterOrders(rows, task) {
  const routeName = normalize(task.routeName);
  const packageName = normalize(task.packageName);
  const startDate = normalize(task.startDate);
  const seen = new Set();

  return rows.filter((row) => {
    if (row["订单状态"] !== "已付款") return false;
    if (row["参团状态"] !== "有效") return false;
    if (!String(row["旅客姓名"] || "").trim()) return false;
    if (!String(row["旅客证件号码"] || "").trim()) return false;
    if (normalize(row["路线"]) !== routeName) return false;
    if (!normalize(row["出行时间"]).startsWith(startDate)) return false;
    if (packageName && normalize(row["套餐"]) !== packageName) return false;

    const idNumber = String(row["旅客证件号码"]).trim();
    if (seen.has(idNumber)) return false;
    seen.add(idNumber);
    return true;
  });
}

function matchRouteInfo(routeRows, orders, task) {
  const packageIds = new Set(orders.map((row) => String(row["套餐编号"] || "").trim()).filter(Boolean));
  let match = routeRows.find((row) => packageIds.has(String(row["套餐编号"] || "").trim()));
  if (match) return match;

  const routeName = normalize(task.routeName);
  const startDate = normalize(task.startDate).replaceAll("-", "/");
  match = routeRows.find((row) => {
    return normalize(row["路线名称"]).includes(routeName) && normalize(row["开始时间"]).startsWith(startDate);
  });
  return match || {};
}

function buildInsurancePayload(task, routeConfig, routeInfo, orders) {
  const startDate = parseDate(task.startDate);
  const endDate = parseDate(task.endDate || task.startDate);
  const durationDays = Number(routeConfig.durationDays || dayDiff(startDate, endDate) + 1 || 1);
  const insuranceStart = addDays(startDate, Number(routeConfig.startOffsetDays || 0));
  const insuranceEnd = addDays(insuranceStart, durationDays - 1);
  const remark = renderTemplate(routeConfig.remarkTemplate || "{routeName} {startDate}", task, routeInfo);

  const travelers = orders.map((row) => ({
    name: String(row["旅客姓名"] || "").trim(),
    gender: String(row["旅客性别"] || "").trim(),
    idType: String(row["旅客证件类型"] || "身份证").trim(),
    idNumber: String(row["旅客证件号码"] || "").trim(),
    birthday: String(row["旅客出生日期"] || "").trim(),
  }));

  return {
    task,
    insurance: {
      category: routeConfig.category || "",
      insurer: routeConfig.insurer || "",
      product: routeConfig.product || "",
      plan: routeConfig.plan || "",
      startDate: formatDate(insuranceStart),
      startTime: routeConfig.startTime || "00:00:00",
      endDate: formatDate(insuranceEnd),
      endTime: routeConfig.endTime || "23:59:59",
      durationDays,
      remark,
      companyCode: routeConfig.companyCode || "",
    },
    routeInfo,
    travelers,
    pasteList: travelers
      .map((item) => [item.name, item.gender, item.idNumber, item.birthday].filter(Boolean).join(" "))
      .join("\n"),
  };
}

function findRouteConfig(routes, routeName) {
  const target = normalize(routeName);
  return routes.find((route) => {
    if (!route.enabled) return false;
    if (normalize(route.routeName) === target) return true;
    const keywords = String(route.keywords || "")
      .split(/[,，、/]/)
      .map(normalize)
      .filter(Boolean);
    return keywords.length > 0 && keywords.every((keyword) => target.includes(keyword));
  });
}

function renderTemplate(template, task, routeInfo) {
  const values = {
    routeName: task.routeName || "",
    packageName: task.packageName || "",
    startDate: task.startDate || "",
    endDate: task.endDate || "",
    routeFullName: routeInfo["路线名称"] || task.routeName || "",
    leader: routeInfo["队长安排"] || "",
  };
  return Object.entries(values).reduce((result, [key, value]) => result.replaceAll(`{${key}}`, value), template);
}

function parseArgs(argv) {
  const parsed = {};
  for (let index = 0; index < argv.length; index += 1) {
    const value = argv[index];
    if (!value.startsWith("--")) continue;
    const key = value.slice(2);
    const next = argv[index + 1];
    parsed[key] = next && !next.startsWith("--") ? next : "true";
    if (parsed[key] === next) index += 1;
  }
  return parsed;
}

function parseDate(value) {
  const [year, month, day] = String(value).slice(0, 10).replaceAll("/", "-").split("-").map(Number);
  return new Date(year, month - 1, day);
}

function addDays(date, days) {
  const next = new Date(date);
  next.setDate(next.getDate() + days);
  return next;
}

function dayDiff(a, b) {
  return Math.round((b.getTime() - a.getTime()) / 86400000);
}

function formatDate(date) {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, "0");
  const day = String(date.getDate()).padStart(2, "0");
  return `${year}-${month}-${day}`;
}

function safeFilename(value) {
  return String(value).replace(/[<>:"/\\|?*\u0000-\u001f]/g, "_");
}

function normalize(value) {
  return String(value || "").trim().toLowerCase();
}

File: automation/clean_data.py
#!/usr/bin/env python3
import argparse
import json
from datetime import datetime, timedelta
from pathlib import Path

import pandas as pd


def main():
    parser = argparse.ArgumentParser(description="清洗订单列表和销转表，生成保险平台投保数据")
    parser.add_argument("--orders", required=True, help="订单列表 xlsx 路径")
    parser.add_argument("--routes", required=True, help="销转表 xlsx 路径")
    parser.add_argument("--config", required=True, help="配置 JSON 路径")
    parser.add_argument("--output-dir", default="automation/outputs", help="输出目录")
    args = parser.parse_args()

    config = json.loads(Path(args.config).read_text(encoding="utf-8"))
    order_df = pd.read_excel(args.orders, dtype=str).fillna("")
    route_df = read_route_export(args.routes)
    output_dir = Path(args.output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    enabled_tasks = [task for task in config["tasks"] if task.get("enabled")]
    results = []

    for task in enabled_tasks:
        route_config = find_route_config(config["routes"], task["routeName"])
        if not route_config:
            raise ValueError(f"未匹配总路线配置：{task['routeName']}")

        task_orders = filter_orders(order_df, task)
        if task_orders.empty:
            results.append({"task": task, "status": "empty", "message": "没有有效已付款旅客"})
            continue

        route_info = match_route_info(route_df, task_orders, task)
        insurance_payload = build_insurance_payload(task, route_config, route_info, task_orders)

        safe_name = safe_filename(f"{task['routeName']}_{task['startDate']}")
        json_path = output_dir / f"{safe_name}.json"
        txt_path = output_dir / f"{safe_name}_粘贴名单.txt"

        json_path.write_text(json.dumps(insurance_payload, ensure_ascii=False, indent=2), encoding="utf-8")
        txt_path.write_text(insurance_payload["pasteList"], encoding="utf-8")

        results.append(
            {
                "task": task,
                "status": "ok",
                "travelerCount": len(insurance_payload["travelers"]),
                "json": str(json_path),
                "pasteList": str(txt_path),
            }
        )

    summary_path = output_dir / "summary.json"
    summary_path.write_text(json.dumps(results, ensure_ascii=False, indent=2), encoding="utf-8")
    print(json.dumps({"summary": str(summary_path), "results": results}, ensure_ascii=False, indent=2))


def read_route_export(path):
    raw = pd.read_excel(path, header=None, dtype=str).fillna("")
    if len(raw) < 2:
        raise ValueError("销转表格式异常：少于两行表头")

    headers = raw.iloc[1].tolist()
    data = raw.iloc[2:].copy()
    data.columns = headers
    return data.fillna("")


def filter_orders(order_df, task):
    route_name = normalize(task["routeName"])
    package_name = normalize(task.get("packageName", ""))
    start_date = normalize(task["startDate"])

    df = order_df.copy()
    df = df[df["订单状态"].eq("已付款")]
    df = df[df["参团状态"].eq("有效")]
    df = df[df["旅客姓名"].str.strip().ne("")]
    df = df[df["旅客证件号码"].str.strip().ne("")]
    df = df[df["路线"].map(normalize).eq(route_name)]
    df = df[df["出行时间"].map(normalize).str.startswith(start_date)]

    if package_name:
      df = df[df["套餐"].map(normalize).eq(package_name)]

    return df.drop_duplicates(subset=["旅客证件号码"], keep="first")


def match_route_info(route_df, task_orders, task):
    package_ids = set(task_orders["套餐编号"].dropna().astype(str))
    route_match = route_df[route_df["套餐编号"].astype(str).isin(package_ids)]

    if route_match.empty:
        route_name = normalize(task["routeName"])
        start_date = normalize(task["startDate"]).replace("-", "/")
        route_match = route_df[
            route_df["路线名称"].map(normalize).str.contains(route_name, regex=False)
            & route_df["开始时间"].map(normalize).str.startswith(start_date)
        ]

    if route_match.empty:
        return {}

    return route_match.iloc[0].to_dict()


def build_insurance_payload(task, route_config, route_info, orders):
    start_date = parse_date(task["startDate"])
    end_date = parse_date(task.get("endDate") or task["startDate"])
    offset_days = int(route_config.get("startOffsetDays") or 0)
    duration_days = int(route_config.get("durationDays") or max((end_date - start_date).days + 1, 1))
    insurance_start = start_date + timedelta(days=offset_days)
    insurance_end = insurance_start + timedelta(days=duration_days - 1)
    remark = render_template(route_config.get("remarkTemplate") or "{routeName} {startDate}", task, route_info)

    travelers = []
    paste_lines = []
    for _, row in orders.iterrows():
        traveler = {
            "name": row["旅客姓名"].strip(),
            "gender": row["旅客性别"].strip(),
            "idType": row["旅客证件类型"].strip() or "身份证",
            "idNumber": row["旅客证件号码"].strip(),
            "birthday": row["旅客出生日期"].strip(),
        }
        travelers.append(traveler)
        paste_lines.append(
            " ".join(
                value
                for value in [traveler["name"], traveler["gender"], traveler["idNumber"], traveler["birthday"]]
                if value
            )
        )

    return {
        "task": task,
        "insurance": {
            "category": route_config.get("category", ""),
            "insurer": route_config.get("insurer", ""),
            "product": route_config.get("product", ""),
            "plan": route_config.get("plan", ""),
            "startDate": insurance_start.strftime("%Y-%m-%d"),
            "startTime": route_config.get("startTime", "00:00:00"),
            "endDate": insurance_end.strftime("%Y-%m-%d"),
            "endTime": route_config.get("endTime", "23:59:59"),
            "durationDays": duration_days,
            "remark": remark,
            "companyCode": route_config.get("companyCode", ""),
        },
        "routeInfo": route_info,
        "travelers": travelers,
        "pasteList": "\n".join(paste_lines),
    }


def find_route_config(routes, route_name):
    target = normalize(route_name)
    for route in routes:
        if not route.get("enabled"):
            continue
        if normalize(route.get("routeName")) == target:
            return route
        keywords = [normalize(item) for item in str(route.get("keywords", "")).replace("，", ",").split(",") if item]
        if keywords and all(keyword in target for keyword in keywords):
            return route
    return None


def render_template(template, task, route_info):
    values = {
        "routeName": task.get("routeName", ""),
        "packageName": task.get("packageName", ""),
        "startDate": task.get("startDate", ""),
        "endDate": task.get("endDate", ""),
        "routeFullName": route_info.get("路线名称", task.get("routeName", "")),
        "leader": route_info.get("队长安排", ""),
    }
    for key, value in values.items():
        template = template.replace("{" + key + "}", str(value))
    return template


def parse_date(value):
    return datetime.strptime(str(value)[:10].replace("/", "-"), "%Y-%m-%d")


def normalize(value):
    return str(value or "").strip().lower()


def safe_filename(value):
    return "".join(char if char.isalnum() or char in "-_" else "_" for char in value)


if __name__ == "__main__":
    main()
